rep-2012.rst

```diff
@@ -0,0 +1,466 @@
+
+Note, we put the service event definitions into the ``msg`` namespace, as this is expected for message interface types in many libraries and tools in the ROS ecosystem.
+
+The service event definitions are generated as part of the ``rosidl`` pipeline [3]_.
+
+Timestamp
+^^^^^^^^^
+Timestamps represent the time at which the event occurred.
+I.e. they are set to the time directly after a request or response is sent or received.
+
+Timestamps shall respect ROS time [4]_.
+This means by default they will be set with wall-time.
+If simulation time is enabled by the node implementing the service or client, then timestamps will get their time from the ``/clock`` topic.
+
+Client ID and sequence number
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Both the client ID and sequence number are provided by the ROS middleware [5]_.
+They can be accessed from ``rcl`` [6]_ when taking a request or response for a service or client respectively.
+
+Example
+^^^^^^^
+
+For example, consider a service ``example_interfaces/srv/AddTwoInts`` defined as follows:
+
+.. code-block::
+
+   int64 a
+   int64 b
+   ---
+   int64 sum
+
+The following service event message definitions are generated when building the ``example_interfaces`` package (comments elided for brevity):
+
+:example_interfaces/msg/AddTwoInts_RequestEvent:
+
+.. code-block::
+
+   uint8 CLIENT_SENT      = 0
+   uint8 SERVICE_RECEIVED = 1
+
+   uint8 request_type
+
+   builtin_interfaces/msg/Time stamp
+
+   unique_identifier_msgs/msg/UUID client_id
+
+   int64 sequence_number
+
+   example_interfaces/srv/AddTwoInts_Request request
+
+:example_interfaces/msg/AddTwoInts_ResponseEvent:
+
+.. code-block::
+
+   uint8 SERVICE_SENT     = 0
+   uint8 CLIENT_RECEIVED  = 1
+
+   uint8 response_type
+
+   builtin_interfaces/msg/Time stamp
+
+   unique_identifier_msgs/msg/UUID client_id
+
+   int64 sequence_number
+
+   example_interfaces/srv/AddTwoInts_Response response
+
+The definition for ``example_interfaces/srv/AddTwoInts_Request`` is,
+
+.. code-block::
+
+   int64 a
+   int64 b
+
+And the definitions for ``example_interfaces/srv/AddTwoInts_Reponse`` is,
+
+.. code-block::
+
+   int64 sum
+
+Configuration
+-------------
+
+Configuration of service introspection features will be done through ROS parameters.
+Each node may offer the following four Boolean parameters:
+
+:publish_client_events:
+  If true, then client "request sent" and "response received" events will be published for all clients created by this node.
+:publish_service_events:
+  If true, then service "request received" and "response sent" events will be published for all services created by this node.
+:client_event_content:
+  If true, then client event messages will include the request and response content.
+  Only applies if ``publish_client_events`` is ``true``.
+:service_event_content:
+  If true, then service event messages will include the request and response content.
+  Only applies if ``publish_service_events`` is ``true``.
```

Comment:
would these be dynamic parameters? Or statically loaded only at startup?

Reply:
I was thinking all parameters would by dynamic. This means event publishers would be created/destroyed as the `publish_*` parameters change and the request/response data would be included in the messages based on the state of the the `*_event_content` parameters.

IMO, this is nice since it let's us turn on introspection temporarily in a live system.